guacamole/src/main/java/org/apache/guacamole/extension/ListenerFactory.java

```diff
@@ -0,0 +1,256 @@
+ *  under the License.
+ */
+
+package org.apache.guacamole.extension;
+
+import org.apache.guacamole.GuacamoleException;
+import org.apache.guacamole.GuacamoleSecurityException;
+import org.apache.guacamole.net.event.AuthenticationFailureEvent;
+import org.apache.guacamole.net.event.AuthenticationSuccessEvent;
+import org.apache.guacamole.net.event.TunnelCloseEvent;
+import org.apache.guacamole.net.event.TunnelConnectEvent;
+import org.apache.guacamole.net.event.listener.*;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * A factory that reflectively instantiates Listener objects for a given
+ * provider class.
+ */
+class ListenerFactory {
+
+    /**
+     * Creates all listeners represented by an instance of the given provider class.
+     * <p>
+     * If a provider class implements the simple Listener interface, that is the
+     * only listener type that will be returned. Otherwise, a list of Listener
+     * objects that adapt the legacy listener interfaces will be returned.
+     *
+     * @param providerClass
+     *      a class that represents a listener provider
+     *
+     * @return
+     *      list of listeners represented by the given provider class
+     */
+    static List<Listener> createListeners(Class<?> providerClass) {
+
+        Object provider = ProviderFactory.newInstance("listener", providerClass);
+
+        if (provider instanceof Listener) {
+            return Collections.singletonList((Listener) provider);
+        }
+
+        return createListenerAdapters(provider);
+
+    }
+
+    @SuppressWarnings("deprecation")
+    private static List<Listener> createListenerAdapters(Object provider) {
+
+        final List<Listener> listeners = new ArrayList<Listener>();
+
+        if (provider instanceof AuthenticationSuccessListener) {
+            listeners.add(new AuthenticationSuccessListenerAdapter(
+                    (AuthenticationSuccessListener) provider));
+        }
+
+        if (provider instanceof AuthenticationFailureListener) {
+            listeners.add(new AuthenticationFailureListenerAdapter(
+                    (AuthenticationFailureListener) provider));
+        }
+
+        if (provider instanceof TunnelConnectListener) {
+            listeners.add(new TunnelConnectListenerAdapter(
+                    (TunnelConnectListener) provider));
+        }
+
+        if (provider instanceof TunnelCloseListener) {
+            listeners.add(new TunnelCloseListenerAdapter(
+                    (TunnelCloseListener) provider));
+        }
+
+        return listeners;
+
+    }
+
+    /**
+     * An adapter the allows an AuthenticationSuccessListener to be used
+     * as an ordinary Listener.
+     */
+    @SuppressWarnings("deprecation")
+    private static class AuthenticationSuccessListenerAdapter implements Listener {
+
+        private final AuthenticationSuccessListener delegate;
+
+        /**
+         * Constructs a new adapter.
```

Comment:
Comments should add information/context which is not inherently present. "Constructs a new adapter" is what any constructor within this class would do, and certainly this constructor has some specific use dealing with its `delegate` parameter.

Consider:

    /**
     * Returns the count.
     *
     * @return
     *     The count.
     */
    public int getCount();

vs:

    /**
     * Returns the number of elements currently in this list.
     *
     * @return
     *     The number of elements in the list.
     */
    public int getCount();